rust-symcrypt/src/hash.rs

```diff
@@ -0,0 +1,186 @@
+    fn drop(&mut self) {
+        unsafe {
+            symcrypt_sys::SymCryptWipe(
+                ptr::addr_of_mut!(self.state) as *mut c_void,
+                mem::size_of_val(&mut self.state) as symcrypt_sys::SIZE_T,
+            )
+        }
+    }
+}
+
+pub fn sha384(data: &[u8], result: &mut [u8; SHA384_RESULT_SIZE]) {
+    unsafe {
+        symcrypt_sys::SymCryptSha384(
+            data.as_ptr(),                      // pbData
+            data.len() as symcrypt_sys::SIZE_T, // cbData
+            result.as_mut_ptr(),                // pbResult
+        );
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::mem::size_of;
+    use symcrypt_sys::{SymCryptSha256Algorithm, SymCryptSha384Algorithm};
+
+    fn check_hash_size(hash: symcrypt_sys::PCSYMCRYPT_HASH) -> symcrypt_sys::SIZE_T {
+        let mut result: symcrypt_sys::SIZE_T = 0;
+        unsafe {
+            result = symcrypt_sys::SymCryptHashStateSize(hash);
+        }
+        result
+    }
+
+    #[test]
+    fn stateless_sha256_hash() {
+        let data = hex::decode("641ec2cf711e").unwrap();
+        let expected: &str = "cfdbd6c9acf9842ce04e8e6a0421838f858559cf22d2ea8a38bd07d5e4692233";
+
+        let mut result: [u8; SHA256_RESULT_SIZE] = [0; SHA256_RESULT_SIZE];
+        sha256(&data, &mut result);
+        assert_eq!(hex::encode(result), expected);
+    }
+
+    #[test]
+    fn state_sha256_hash() {
+        let data = hex::decode("").unwrap();
+        let expected: &str = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+
+        let mut sha_test = Sha256State::new();
+        let mut result: [u8; SHA256_RESULT_SIZE] = [0; SHA256_RESULT_SIZE];
+        Sha256State::append(&mut sha_test, &data);
+        Sha256State::result(&mut sha_test, &mut result);
+        assert_eq!(hex::encode(result), expected);
+    }
+
+    #[test]
+    fn stateless_sha384_hash() {
+        let data = hex::decode("").unwrap();
+        let expected: &str = "38b060a751ac96384cd9327eb1b1e36a21fdb71114be07434c0cc7bf63f6e1da274edebfe76f65fbd51ad2f14898b95b";
+
+        let mut result: [u8; SHA384_RESULT_SIZE] = [0; SHA384_RESULT_SIZE];
+        sha384(&data, &mut result);
+        assert_eq!(hex::encode(result), expected);
+    }
+
+    #[test]
+    fn state_sha384_hash() {
+        let data = hex::decode("f268267bfb73d5417ac2bc4a5c64").unwrap();
+        let expected: &str = "6f246b1f839e73e585c6356c01e9878ff09e9904244ed0914edb4dc7dbe9ceef3f4695988d521d14d30ee40b84a4c3c8";
+
+        let mut sha_test = Sha384State::new();
+        let mut result: [u8; SHA384_RESULT_SIZE] = [0; SHA384_RESULT_SIZE];
+        Sha384State::append(&mut sha_test, &data);
+        Sha384State::result(&mut sha_test, &mut result);
+        assert_eq!(hex::encode(result), expected);
+    }
+
+    #[test]
+    fn check_state_size() {
+        unsafe {
+            assert_eq!(
+                check_hash_size(SymCryptSha256Algorithm),
+                size_of::<Sha256State>() as u64
```

Comment:
You can add `#[repr(transparent)]` on `Sha256State` and `Sha384State` if you want to provide a stronger guarantee that the sizes match.

Reply:
I read over some documentation on `#[repr(transparent)]`, from what I understand it enforces the Rust struct has the same memory layout and ABI as the corresponding foreign struct. When Phil and I stepped through the debugger that already seemed to be the case. Do you know if there would be any added benefit for adding the `#[repr(transparent)]`? I think I am more inclined to just trust the rust compiler

Reply:
Bindgen doesn't  add `#[repr(transparent)]` but it does provide a self test which taking a look should be sufficient in testing the layout is the same. [bindgen_test_layout__SYMCRYPT_SHA256_STATE](https://github.com/nnmkhang/rust-symcrypt-sys/blob/main/symcrypt-sys/src/symcrypt_bindings.rs#L158)

Reply:
Removed code since the bindgen tests cover the size and layout tests.